cmd/juju/application/refresher/refresher.go

```diff
@@ -0,0 +1,268 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package refresher
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/juju/charm/v8"
+	"github.com/juju/charmrepo/v6"
+	jujuclock "github.com/juju/clock"
+	"github.com/juju/errors"
+	"github.com/juju/featureflag"
+
+	"github.com/juju/juju/cmd/juju/application/store"
+	"github.com/juju/juju/cmd/juju/application/utils"
+	corecharm "github.com/juju/juju/core/charm"
+	"github.com/juju/juju/feature"
+)
+
+// ErrExhausted reveals if a refresher was exhausted in it's task. If so, then
+// it is expected to attempt another refresher.
+var ErrExhausted = errors.Errorf("exhausted")
+
+// RefresherDependencies are required for any deployer to be run.
+type RefresherDependencies struct {
+	Authorizer    store.MacaroonGetter
+	CharmAdder    store.CharmAdder
+	CharmResolver CharmResolver
+}
+
+// RefresherConfig is the data required to choose an refresher and then run the
+// PrepareAndUpgrade.
+type RefresherConfig struct {
+	ApplicationName string
+	CharmURL        *charm.URL
+	CharmRef        string
+	Channel         corecharm.Channel
+	DeployedSeries  string
+	Force           bool
+	ForceSeries     bool
+}
+
+type RefresherFn = func(RefresherConfig) (Refresher, error)
+
+type factory struct {
+	refreshers []RefresherFn
+	clock      jujuclock.Clock
+}
+
+// NewRefresherFactory returns a factory setup with the API and
+// function dependencies required by every refresher.
+func NewRefresherFactory(deps RefresherDependencies) RefresherFactory {
+	d := &factory{
+		clock: jujuclock.WallClock,
+	}
+	d.refreshers = []RefresherFn{
+		d.maybeReadLocal(deps.CharmAdder, defaultCharmRepo{}),
+		d.maybeCharmStore(deps.Authorizer, deps.CharmAdder, deps.CharmResolver),
+	}
+	return d
+}
+
+// GetRefresher returns the correct refresher to use based on the cfg provided.
```

Comment:
Comment is wrong. There are some other exported methods in this patch missing comments too.